solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1299,6 +1385,47 @@ public DocList getDocList(Query query, List<Query> filterList, Sort lsort, int o
   public static final int GET_DOCLIST = 0x02; // get the documents actually returned in a response
   public static final int GET_SCORES = 0x01;
 
+  private static boolean isConstantScoreQuery(Query q) {
```

Comment:
I've seen a need for this elsewhere.  Let's make it handle other cases too (e.g. MatchAllDocsQuery, MatchNoDocsQuery, "Filter" (soon to be DocSetQuery)) and be recursive to the cases below (hey why not).  This could move to QueryUtils rather than having the massive SolrIndexSearcher grow

Reply:
:+1: done in dfff5f792a52ff4c8ed87d4495745340c7af10e6. Lmk if this is _too_ exhaustive, or if I've missed anything?

Reply:
>I've seen a need for this elsewhere

I looked for other places in the codebase where this could be used, and I didn't immediately find anything. Perhaps I missed something? It does indeed seem general-purpose useful (even if only otherwise leveraged from 3rd-party plugins).

Reply:
`org.apache.solr.search.QueryUtils#combineQueryAndFilter` wraps a ConstantScoringQuery but it could avoid it.

Reply:
FYI, I had to revert the above change; see: c9ccf5c8be98a169742ffc924dda24aa8d3bc4e0

Reply:
was that uncovered by a test (which)?

Reply:
Yes (sorry, I mentioned it in the commit):
```
gradlew :solr:core:test --tests "org.apache.solr.search.TestPseudoReturnFields.testMultiValued"
```

Reply:
coincidentally enough I think this may be somewhat related to your comment on [PR #529](https://github.com/apache/solr/pull/529#discussion_r803130966)? Given that this issue has some TODOs related to SOLR-12336 anyway, I'll probably just wait to merge this until SOLR-12336 lands on main and 9_x.

Reply:
Thanks; I'll look into this after that PR is merged.